File: commons/src/main/java/com/tangula/android/base/TglService.java
package com.tangula.android.base;

import android.app.Activity;
import android.app.Service;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;

import com.tangula.android.utils.ApplicationUtils;
import com.tangula.android.utils.LogUt;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.functions.Consumer;
import kotlin.Unit;

public abstract class TglService<S extends IServiceBinder> extends Service {

    private static final int BIND_FLAG_AUTO_CRATE_AND_ALWAY_START = Context.BIND_WAIVE_PRIORITY | Context.BIND_AUTO_CREATE;


    private static final int BIND_FLAG_AUTO_CRATE_AND_ALWAY_START_IN_BG = Context.BIND_WAIVE_PRIORITY | Context.BIND_AUTO_CREATE | Context.BIND_NOT_FOREGROUND;

    protected static final List<IServiceBinder> BINDERS = new ArrayList<>(5);


    public static <S extends IServiceBinder> void startService(Activity act, Class<? extends TglService<S>> serviceType) {
        if (act != null) {
            act.startService(new Intent(act, serviceType));
        } else {
            ApplicationUtils.APP.startService(new Intent(ApplicationUtils.APP, serviceType));
        }

    }


    public static <S extends IServiceBinder> ServiceConnection bind2Service(Context ctx,
                                                                            Class<? extends TglService<S>> serviceType,
                                                                            int flags,
                                                                            final Consumer<S> afterConnected,
                                                                            final Consumer<ComponentName> onDisConnect,
                                                                            final Runnable onBindFail) {
        Intent intent = new Intent(ctx, serviceType);
        ServiceConnection conn = new ServiceConnection() {
            private S binder = null;

            @Override
            public void onServiceConnected(ComponentName name, IBinder service) {
                if (binder != null) {
                    try {
                        afterConnected.accept(binder);
                    } catch (Exception e) {
                        LogUt.e(e);
                    }
                } else {
                    if (onBindFail != null) {
                        onBindFail.run();
                    }
                }
            }

            @Override
            public void onServiceDisconnected(ComponentName name) {

                if (onDisConnect != null) {
                    try {
                        onDisConnect.accept(name);
                    } catch (Exception e) {
                        LogUt.e(e);
                    }
                }

                if (binder != null) {
                    binder.close();
                    BINDERS.remove(binder);
                }

            }

        };
        ctx.bindService(intent, conn, flags);
        return conn;
    }

    public static <I extends IServiceBinder> ServiceConnection bind2NormalService(Context ctx, Class<? extends TglService<I>> serviceType, final Consumer<I> afterConnected,
                                                                                  final Consumer<ComponentName> onDisConnect,
                                                                                  final Runnable onBindFail) {
        return bind2Service(ctx, serviceType, BIND_FLAG_AUTO_CRATE_AND_ALWAY_START, afterConnected, onDisConnect, onBindFail);
    }

    public static <I extends IServiceBinder> ServiceConnection bind2BackgroundService(Context ctx, Class<? extends TglService<I>> serviceType, final Consumer<I> afterConnected,
                                                                                      final Consumer<ComponentName> onDisConnect,
                                                                                      final Runnable onBindFail) {
        return bind2Service(ctx, serviceType, BIND_FLAG_AUTO_CRATE_AND_ALWAY_START_IN_BG, afterConnected, onDisConnect, onBindFail);
    }


    /**
     * 是否显示在前台.
     */
    private ForegroundServiceMetaInfo foreGroundInfo = null;

    private boolean isSingleton = true;

    /**
     * 服务实现对象.
     */
    private S instance = null;

    public IBinder onBind(Intent intent) {
        S res;
        if (isSingleton) {
            if (instance == null) {
                synchronized (this) {
                    if (instance == null) {
                        instance = createBinder(intent, -1, -1);
                    }
                }
            }
            res = instance;
        }else {
            res = createBinder(intent, -2, -2);
        }
        if (res != null) {
            res.onStart();
        }
        return res;
    }

    public boolean onUnbind(Intent intent){
        if (isSingleton) {
            synchronized (this) {
                if(instance!=null){
                    instance.close();
                }
            }
        }
        return super.onUnbind(intent);
    }

    public abstract S onCreateBinder(Intent intent, int flags, int startId);

    public S createBinder(Intent intent, int flags, int startId)

    {
        S res = onCreateBinder(intent, flags, startId);
        if (res != null) {
            res.setServiceInstance(this);
            BINDERS.add(res);
        }
        return res;
    }

    public void onCreate() {
        super.onCreate();
        if (isSingleton) {
            instance = createBinder(null, -1, -1);
        }
    }

    public int onStartCommand(Intent intent, int flags, int startId) {
        if(foreGroundInfo!=null) {
            startForeground(foreGroundInfo.getServiceId(), new ForeServiceNotificationBuilder().buildNotification(this.getApplicationContext(), foreGroundInfo.getIconRes(), foreGroundInfo.getChannelId(), foreGroundInfo.getChannelName()));
        }

        if (isSingleton) {

            if(instance!=null){
                return instance.onStart();
            }else{
                return START_STICKY;
            }

        } else {
            return createBinder(null, flags, startId).onStart();
        }
    }

    public void onLowMemory() {
        super.onLowMemory();
        for(IServiceBinder it:BINDERS){
            it.onLowMemory();
        }
    }

    public void onTrimMemory(int level) {
        super.onTrimMemory(level);
        for(IServiceBinder it:BINDERS){
            it.onTrimMemory(level);
        }
    }

    public void onDestroy() {
        super.onDestroy();
        for(IServiceBinder it:BINDERS){
            it.onDestroy();
        }
    }

}

File: commons/src/main/java/com/tangula/android/base/TglService.kt
package com.tangula.android.base

import android.annotation.SuppressLint
import android.app.*
import android.content.*
import android.os.Binder
import android.os.IBinder
import android.os.IInterface
import android.support.annotation.DrawableRes
import android.util.Log
import com.tangula.android.utils.ApplicationUtils
import com.tangula.android.utils.SdkVersionDecides

/**
 * 前台服务元数据.
 */
data class ForegroundServiceMetaInfo(val serviceId: Int, val iconRes: Int, val channelId: String, val channelName: String?)


/**
 * 为前台服务通知构建器.
 */
class ForeServiceNotificationBuilder {
    @SuppressLint("NewApi")
    fun buildNotification(ctx: Context, @DrawableRes iconResId: Int, channelId: String, channelName: String?): Notification {

        val manager: NotificationManager = ctx.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val builder: Notification.Builder

        builder = if (SdkVersionDecides.afterSdk26A8d0()) {
            val channel = NotificationChannel(channelId, channelName, NotificationManager.IMPORTANCE_HIGH)
            manager.createNotificationChannel(channel)
            Notification.Builder(ctx, channelId)
        } else {
            Notification.Builder(ctx)
        }

        SdkVersionDecides.afterSdk20A4d4W {
            builder.setLocalOnly(true)
        }

        SdkVersionDecides.afterSdk17A4d2 {
            builder.setShowWhen(false)
        }

        builder.setContentTitle(channelId).setContentText(channelName)
                .setSmallIcon(iconResId).setAutoCancel(false)

        return builder.build()
    }
}

interface IServiceBinder:IInterface,IBinder{

    var serviceInstance : Service?

    /**
     * 每次bind或者start的时候执行.
     * @return 参见:[Service.onStartCommand]
     */
    fun onStart():Int

    /**
     *  服务被unbind的时候被执行.
     */
    fun close()

    fun onLowMemory()

    fun onTrimMemory(level: Int)

    fun onDestroy()
}

open class TglLocalServiceBinder:Binder(), IServiceBinder{

    override var serviceInstance : Service? = null

    override fun onStart(): Int {
        return Service.START_STICKY
    }

    override fun close() {
    }

    override fun onLowMemory() {
    }

    override fun onTrimMemory(level: Int) {
    }

    override fun onDestroy() {
    }

    override fun asBinder(): IBinder {
        return this
    }
}



abstract class TglService<S: IServiceBinder> : Service() {

    companion object {

        val BIND_FLAG_AUTO_CRATE_AND_ALWAY_START = Context.BIND_WAIVE_PRIORITY or Context.BIND_AUTO_CREATE

        val BIND_FLAG_AUTO_CRATE_AND_ALWAY_START_IN_BG = Context.BIND_WAIVE_PRIORITY or Context.BIND_AUTO_CREATE or Context.BIND_NOT_FOREGROUND

        protected val BINDERS:MutableList<IServiceBinder> = mutableListOf()

        fun <S: IServiceBinder> startService(act: Activity?, serviceType:Class<out TglService<S>>) {
            if (act is Activity) {
                    act.startService(Intent(act, serviceType))
            }else{
                ApplicationUtils.APP.startService(Intent(ApplicationUtils.APP, serviceType))
            }

        }

        fun <S:IServiceBinder> bind2Service(ctx:Context, serviceType: Class<out TglService<S>>, flags:Int, afterConnected:((S)->Unit)?, onDisConnect:((ComponentName?)->Unit)?, onBindFail:(()->Unit)?): ServiceConnection{
            val intent = Intent(ctx, serviceType)
            val conn = object: ServiceConnection {
                var binder: S? = null
                override fun onServiceDisconnected(name: ComponentName?) {
                    onDisConnect?.also { it(name) }
                    binder?.also {
                        it.close()
                        BINDERS.remove(it)
                    }
                }
                override fun onServiceConnected(name: ComponentName?, binder: IBinder?) {
                    if(binder is IServiceBinder){
                        afterConnected?.also { it(binder!! as S) }
                    }else{
                        onBindFail?.also { it() }
                    }
                }
            }
            ctx.bindService(intent, conn,flags)
            return conn
        }

        fun <I:IServiceBinder>  bind2NormalService(ctx:Context, serviceType: Class<out TglService<I>>,afterConnected:((I)->Unit)?, onDisConnect:((ComponentName?)->Unit)?, onBindFail:(()->Unit)?):ServiceConnection{
            return bind2Service(ctx, serviceType, BIND_FLAG_AUTO_CRATE_AND_ALWAY_START,afterConnected , onDisConnect, onBindFail)
        }

        fun <I:IServiceBinder>  bind2BackgroundService(ctx:Context, serviceType: Class<out TglService<I>>,afterConnected:((I)->Unit)?, onDisConnect:((ComponentName?)->Unit)?, onBindFail:(()->Unit)?):ServiceConnection{
            return bind2Service(ctx, serviceType, BIND_FLAG_AUTO_CRATE_AND_ALWAY_START_IN_BG,afterConnected , onDisConnect, onBindFail)
        }

    }


    /**
     * 是否显示在前台.
     */
    var foreGroundInfo: ForegroundServiceMetaInfo? = null

    var isSingleton = true

    /**
     * 服务实现对象.
     */
    var instance: S? = null

    override fun onBind(intent: Intent?): IBinder {
        if(isSingleton){
            if(instance == null){
                synchronized(this) {
                     if(instance==null) {
                         instance = createBinder(intent, -1, -1)
                     }
                }
            }

            return instance!!.also { it.onStart() }
        }
        return createBinder(intent, -2, -2).also { it.onStart() }
    }

    override fun onUnbind(intent: Intent?): Boolean {
        if(isSingleton){
            synchronized(this){
                instance?.also {
                    it.close()
                }
            }
        }
        return super.onUnbind(intent)
    }

    abstract fun onCreateBinder(intent:Intent?,flags:Int, startId:Int):S

    fun createBinder(intent: Intent?, flags: Int, startId: Int): S{
        Log.v("console","TglService::createBinder")
        return onCreateBinder(intent, flags, startId).also {
            it.serviceInstance =  this
            BINDERS.add(it)
        }
    }

    override fun onCreate() {
        Log.v("console","TglService::onCreate")
        super.onCreate()
        if(isSingleton) {
            instance = createBinder(null, -1, -1)
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        Log.v("console","TglService::onStartCommand")
        foreGroundInfo?.also {
            startForeground(it.serviceId, ForeServiceNotificationBuilder().buildNotification(applicationContext, it.iconRes, it.channelId, it.channelName))
        }

        if(isSingleton){
            return instance?.onStart()?: START_STICKY
        }else{
            return createBinder(null, flags, startId).onStart()
        }
    }

    override fun onLowMemory() {
        super.onLowMemory()
        BINDERS.forEach{
            it.onLowMemory()
        }
    }

    override fun onTrimMemory(level: Int) {
        super.onTrimMemory(level)
        BINDERS.forEach{
            it.onTrimMemory(level)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        BINDERS.forEach{
            it.onDestroy()
        }
    }

}
